app/src/components/widgets/charts/launchStatisticsChart/launchStatisticsChart.jsx

```diff
@@ -0,0 +1,427 @@
+    this.chart = chart;
+    this.node = element;
+
+    if (this.props.isPreview) {
+      return;
+    }
+
+    this.isCustomTooltipNeeded() ? this.onCustomChartCreated() : this.onDefaultChartCreated();
+  };
+
+  onDefaultChartCreated = () => {
+    this.node.addEventListener('mousemove', this.getCoords);
+  };
+
+  onCustomChartCreated = () => {
+    this.interactElems = d3.selectAll(
+      this.node.querySelectorAll(this.isSingleColumn() ? '.c3-circle' : '.c3-area'),
+    );
+
+    if (this.isSingleColumn()) {
+      // eslint-disable-next-line
+      this.interactElems._groups[0].forEach(function (item) {
+        if (!item.value) {
+          d3.select(this).style('display', 'none');
+        }
+      });
+    }
+
+    this.createInteractiveTooltip();
+  };
+
+  onMouseOut = () => {
+    this.chart.revert();
+  };
+
+  onMouseOver = (id) => {
+    this.chart.focus(id);
+  };
+
+  onClick = (id) => {
+    this.chart.toggle(id);
+  };
+
+  onChartClick = (data) => {
+    if (this.isTimeLine) {
+      return; // TODO: do it when filters will be available ob launches page
+    }
+    const { widget, getDefectLink, getStatisticsLink, defectTypes } = this.props;
+    const nameConfig = getItemNameConfig(data.id);
+    const id = widget.content.result[data.index].id;
+    const defaultParams = this.getDefaultLinkParams(id);
+    const locators = getDefectTypeLocators(nameConfig, defectTypes);
+
+    const link = locators
+      ? getDefectLink({ defects: locators, itemId: id })
+      : getStatisticsLink({ statuses: this.getLinkParametersStatuses(nameConfig) });
+    this.props.redirect(Object.assign(link, defaultParams));
+  };
+
+  onItemMouseOver = () => this.tooltip.style('display', 'block');
+
+  onItemMouseOut = () => this.tooltip.style('display', 'none');
+
+  onItemMouseMove = (data) => {
+    const rectWidth = this.node.querySelectorAll('.c3-event-rect')[0].getAttribute('width');
+    const currentMousePosition = d3.mouse(this.chart.element);
+    const itemWidth = rectWidth / data.values.length;
+    const dataIndex = Math.trunc((currentMousePosition[0] - CHART_OFFSET) / itemWidth);
+    this.selectedLaunchData = data.values.find((item) => item.index === dataIndex);
+    this.tooltip
+      .html(() =>
+        this.renderContents([this.selectedLaunchData], null, null, (id) => this.colors[id]),
+      )
+      .style('left', `${currentMousePosition[0] - 108}px`)
+      .style('top', `${currentMousePosition[1] - 72}px`);
+  };
+
+  getLinkParametersStatuses = ({ defectType }) => {
+    if (defectType === TOTAL_KEY) {
+      return [PASSED, FAILED, SKIPPED, INTERRUPTED];
+    }
+    return [defectType.toUpperCase()];
+  };
+
+  getDefaultLinkParams = (testItemIds) => ({
+    payload: {
+      projectId: this.props.project,
+      filterId: 'all',
+      testItemIds,
+    },
+    type: TEST_ITEM_PAGE,
+  });
+
+  getCoords = ({ pageX, pageY }) => {
+    this.x = pageX;
+    this.y = pageY;
+  };
+
+  getConfig = () => {
```

Comment:
Is it possible to split this function into a few smaller ones? It's just difficult to understand what happens here.

Reply:
Done